plugins/catalog-backend/src/ingestion/processors/ldap/vendors.ts

```diff
@@ -0,0 +1,103 @@
+/*
+ * Copyright 2021 Spotify AB
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { SearchEntry } from 'ldapjs';
+
+/**
+ * An LDAP Vendor handles unique nuances between different vendors.
+ */
+export type LdapVendor = {
+  /**
+   * The attribute name that holds the distinguished name (DN) for an entry.
+   */
+  dnAttributeName: string;
+  /**
+   * The attribute name that holds a universal unique identifier for an entry.
+   */
+  uuidAttributeName: string;
+  /**
+   * Decode ldap entry values for a given attribute name to their string representation.
+   *
+   * @param entry The ldap entry
+   * @param name The attribute to decode
+   */
+  decodeStringAttribute: (entry: SearchEntry, name: string) => string[];
+};
+
+export const DefaultLdapVendor: LdapVendor = {
+  dnAttributeName: 'entryDN',
+  uuidAttributeName: 'entryUUID',
+  decodeStringAttribute: (entry, name) => {
+    return decode(entry, name, value => {
+      return value.toString();
+    });
+  },
+};
+
+export const ActiveDirectoryVendor: LdapVendor = {
+  dnAttributeName: 'distinguishedName',
+  uuidAttributeName: 'objectGUID',
+  decodeStringAttribute: (entry, name) => {
+    const decoder = (value: string | Buffer) => {
+      if (name === ActiveDirectoryVendor.uuidAttributeName) {
+        return formatGUID(value);
+      }
+      return value.toString();
+    };
+    return decode(entry, name, decoder);
+  },
+};
+
+// Decode an attribute to a consumer
+function decode(
+  entry: SearchEntry,
+  attributeName: string,
+  decoder: (value: string | Buffer) => string,
+): string[] {
+  const values = entry.raw[attributeName];
+  if (Array.isArray(values)) {
+    return (values as Buffer[]).map(v => {
+      return decoder(v);
+    });
+  } else if (values) {
+    return [decoder(values)];
+  }
+  return [];
+}
+
+// Formats a Microsoft Active Directory binary-encoded uuid to a readable string
+// See https://github.com/ldapjs/node-ldapjs/issues/297#issuecomment-137765214
+function formatGUID(objectGUID: string | Buffer | Buffer[]): string {
+  let data: Buffer | Buffer[];
```

Comment:
Not sure about what would actually happen here if the argument was a `Buffer[]`. Don't we need to do like `data = [objectGUID].flat()` to  make sure it's always an array, and check that there was exactly one item? What does even `Buffer.toString(16)`, rather than `Buffer[i].toString()`, mean in this context (which is what would happen if we got an array without some change to the code as proposed)?

Reply:
The argument will never be a Buffer[] since each value in an array will be decoded separately. As you can see in the logic https://github.com/backstage/backstage/pull/5084/files#diff-bc71af02c4ca435eb77be9dca3182a9dea2d8939426c14a48ce4bb6996e67303R71-R74